File: ffcontents/ffcontents/ContentsBuilder.m
//
//  ContentsBuilder.m
//  ffcontents
//
//  Created by Peter Kollath on 13/11/14.
//  Copyright (c) 2014 Peter Kollath. All rights reserved.
//

#import "ContentsBuilder.h"
#import "FlatFileUtils.h"

@implementation ContentsBuilder

-(id)init
{
    self = [super init];
    if (self)
    {
        self.whiteSpaces = [NSCharacterSet whitespaceAndNewlineCharacterSet];
        self.contentTaggedItems = [[NSMutableDictionary alloc] init];
        self.stpdefs = [[NSMutableDictionary alloc] init];
        self.contentArray = [[NSMutableArray alloc] init];

    }
    return self;
}


-(BOOL)validate
{
    return (self.inputFile.length > 0 && self.outputDir.length > 0 && self.levelFile.length > 0);
}

-(void)loadLevels
{
    self.levels = [NSMutableDictionary new];
    NSString * str = [NSString stringWithContentsOfFile:self.levelFile
                                               encoding:NSUTF8StringEncoding
                                                  error:NULL];
    NSArray * lines = [str componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];
    for (NSString * line in lines)
    {
        NSArray * parts = [line componentsSeparatedByString:@"\t"];
        if (parts.count == 3)
        {
            [[self levels] setObject:[NSNumber numberWithInt:[(NSString *)parts[0] intValue]]
                              forKey:parts[2]];
        }
    }
}

-(void)process
{
    [self loadLevels];
    
    int count = 0;
    int len = 0;
    FILE * i = fopen([self.inputFile UTF8String], "rt");
    
    self->outputFile = fopen([[self.outputDir stringByAppendingPathComponent:@"contents.txt"] UTF8String], "wt");
    self->textFile = fopen([[self.outputDir stringByAppendingPathComponent:@"texts-b.txt"] UTF8String], "wt");

    char *c = malloc(200000);
    while((c = fgets(c, 200000, i)) != NULL)
    {
        len = (int)strlen(c);
        if (len > 0 && c[len-1] == '\n')
            c[len-1] = '\0';
        NSString * line = [NSString stringWithUTF8String:c];
        NSArray * part = [line componentsSeparatedByString:@"\t"];
        NSString * text = part[1];
        NSString * levelName = part[2];
        NSString * styleName = part[3];
        NSInteger recId = [part[0] integerValue];

        NSMutableDictionary * dict = [NSMutableDictionary new];
        [dict setObject:[NSNumber numberWithInteger:recId] forKey:@"RECORDID"];
        [dict setObject:part[2] forKey:@"LEVELNAME"];
        [dict setObject:part[3] forKey:@"STYLENAME"];
        [dict setObject:@"plain" forKey:@"curr_flow"];
        [dict setObject:[NSMutableString new] forKey:@"plain"];
        [self setDict:dict withText:text];
        
        fprintf(self->textFile, "%ld\t%s\t%s\t%s\n", recId,
                [(NSString *)[dict valueForKey:@"plain"] UTF8String],
                (levelName ? [levelName UTF8String] : ""),
                (styleName ? [styleName UTF8String] : ""));
        
        [self processText:dict];
        count ++;
        
        if (count % 1000 == 0) {
            NSString * str = [NSString stringWithFormat:@"processed %d lines", count];
            printf("%s", str.UTF8String);
            for(int i = 0; i < str.length; i++)
            {
                printf("\b");
            }
            fflush(stdout);
        }
    }
    
    [self saveContents];
    
    fclose(i);
    fclose(self->outputFile);
    fclose(self->textFile);
}


-(void)processText:(NSDictionary *)dict
{
    NSString * text = [dict valueForKey:@"plain"];
    NSInteger recId = [[dict valueForKey:@"RECORDID"] integerValue];
    NSString * style = [dict valueForKey:@"STYLENAME"];
    NSString * level = [dict valueForKey:@"LEVELNAME"];
    
    NSDictionary * contentBuilding = [self.lastContentItemInserted objectForKey:@"STPDEF"];
    if (contentBuilding != nil)
    {
        NSString * hook = [contentBuilding objectForKey:level];
        
        if (hook == nil)
        {
            hook = [contentBuilding objectForKey:style];
        }
        
        if (hook != nil)
        {
            NSString * plainText = [self makeContentTextFromRecord:text];
            NSString * addedText = [NSString stringWithFormat:@"<STP:%@>%@", hook, plainText];
            NSMutableString * target = [self.lastContentItemInserted objectForKey:@"subtext"];
            [target appendString:addedText];
        }
    }
    
    // save record to contents
    [self saveRecordToContents:dict];

}

-(NSString *)makeContentTextFromRecord:(NSString *)plainText
{
    NSString * contentTextCandidate = plainText;
    
    NSRange bh = [plainText rangeOfString:@"<BH>"];
    NSRange eh = [plainText rangeOfString:@"<EH>"];
    NSRange rng;
    
    if (bh.location != NSNotFound)
    {
        if (eh.location != NSNotFound && eh.location > bh.location) {
            rng = NSMakeRange(bh.location + bh.length, eh.location - bh.location - bh.length);
        } else {
            rng = NSMakeRange(bh.location + bh.length, plainText.length - bh.location - bh.length);
        }
        contentTextCandidate = [[contentTextCandidate substringWithRange:rng] stringByTrimmingCharactersInSet:self.whiteSpaces];
    }
    else {
        contentTextCandidate = [contentTextCandidate stringByTrimmingCharactersInSet:self.whiteSpaces];
    }
    
    return [FlatFileUtils removeTagsAndNotes:contentTextCandidate];
}

-(NSUInteger)getLevelIndex:(NSString *)levelName
{
    if (levelName == nil || [levelName length] == 0)
        return NSNotFound;
    
    NSNumber * index = [self.levels objectForKey:levelName];
    if (index != nil) {
        return [index unsignedIntegerValue];
    }

    return NSNotFound;
}

-(BOOL)checkSupressedContentLevel:(NSUInteger)nLevel
{
    for (int i = kContStripMax - 1; i >= 0; i--)
    {
        if (contStrips[i] != nil && [contStrips[i] valueForKey:@"STPLAST"] != nil)
        {
            NSString * str = [contStrips[i] valueForKey:@"STPLAST"];
            if (str != nil) {
                NSUInteger level = [self getLevelIndex:str];
                // first STPLAST found is actually last STPLAST defined
                // so last STPLAST overrides all previously defined
                // therefore return after evaluating level index
                // :: current level (nLevel) must not be higher than level of found STPLAST
                // :: in order to write content item into contents
                return (level < nLevel);
            }
        }
    }
    
    return NO;
}

-(void)setDict:(NSMutableDictionary *)dict withText:(NSString *)str
{
    if (str == nil)
        return;
    
    int start = 0;
    int end = 0;
    int status = 0;
    BOOL foundTag = NO;
    BOOL foundChar = NO;
    BOOL foundCharSpec = NO;
    
    for(int i = 0; i < [str length]; i++)
    {
        if (status == 0) {
            if ([str characterAtIndex:i] == '<') {
                status = 1;
            } else {
                foundChar = YES;
            }
        } else if (status == 1) {
            if ([str characterAtIndex:i] == '<') {
                status = 0;
                foundCharSpec = YES;
            } else {
                start = i-1;
                status = 2;
            }
        } else if (status == 2) {
            if ([str characterAtIndex:i] == '>') {
                end = i;
                foundTag = YES;
                status = 0;
            } else if ([str characterAtIndex:i] == '"') {
                status = 3;
            }
        } else if (status == 3) {
            if ([str characterAtIndex:i] == '"') {
                status = 4;
            }
        } else if (status == 4) {
            if ([str characterAtIndex:i] == '"') {
                status = 3;
            } else if ([str characterAtIndex:i] == '>') {
                end = i;
                foundTag = YES;
                status = 0;
            }
        }
        
        if (foundTag == YES) {
            NSRange foundRange = NSMakeRange(start, end - start + 1);
            NSString * extractedTag = [str substringWithRange:foundRange];
            FlatFileTagString * tag = [FlatFileTagString new];
            [tag clear];
            [tag appendString:extractedTag];
            foundTag = NO;
            NSArray * arr = [tag createArray];

            if ([arr[0] isEqualToString:@"FLOW"])
            {
                [dict setObject:arr[2] forKey:@"curr_flow"];
            }
            else if ([arr[0] isEqualToString:@"STPLAST"])
            {
                if ([arr count] == 1)
                {
                    [dict setValue:@"" forKey:@"STPLAST"];
                }
                else if ([arr count] == 3)
                {
                    NSString * paraName = [arr objectAtIndex:2];
                    NSString * safeString = [FlatFileString stringToSafe:paraName tag:@"LE"];
                    [dict setValue:safeString forKey:@"STPLAST"];
                }
            }
            else if ([arr[0] isEqualToString:@"CTDEF"]) {
                if ([arr count] > 2)
                {
                    [dict setValue:[arr objectAtIndex:2] forKey:@"ctdef"];
                }
            }
            else if ([arr[0] isEqualToString:@"CTUSE"]) {
                if ([arr count] > 2)
                {
                    [dict setValue:[arr objectAtIndex:2] forKey:@"ctuse"];
                }
            } else if ([arr[0] isEqualToString:@"STPDEF"]) {
                if ([arr count] == 1)
                {
                    [self.stpdefs removeAllObjects];
                }
                else if ([arr count] == 5)
                {
                    [self.stpdefs removeObjectForKey:[arr objectAtIndex:4]];
                }
                else if ([arr count] == 13)
                {
                    // tagArr[2] must be equal to LV
                    // tagArr[10] must be equal to STP
                    NSString * targetString = [FlatFileString stringToSafe:arr[4] tag:@"LE"];
                    NSString * hookString = [FlatFileString stringToSafe:[arr objectAtIndex:8]
                                                                     tag:[arr objectAtIndex:6]];
                    NSMutableDictionary * targetDict = nil;
                    
                    targetDict = [self.stpdefs objectForKey:targetString];
                    if (targetDict == nil)
                    {
                        targetDict = [[NSMutableDictionary alloc] init];
                        [self.stpdefs setObject:targetDict forKey:targetString];
                    }
                    
                    [targetDict setObject:[arr objectAtIndex:12] forKey:hookString];
                }
            }
            else
            {
                NSMutableString * ms = [self currentFlowText:dict];
                [ms appendString:extractedTag];
            }
        }
        else if (foundChar == YES)
        {
            NSMutableString * ms = [self currentFlowText:dict];
            [ms appendFormat:@"%C", [str characterAtIndex:i]];
            foundChar = NO;
        }
        else if (foundCharSpec == YES)
        {
            NSMutableString * ms = [self currentFlowText:dict];
            [ms appendFormat:@"<<"];
            foundCharSpec = NO;
        }
    }
    
    return;

}

-(NSMutableString *)currentFlowText:(NSMutableDictionary *)dict
{
    NSMutableString * ms = [dict objectForKey:[dict objectForKey:@"curr_flow"]];
    if (ms == nil) {
        ms = [NSMutableString new];
        [dict setObject:ms forKey:[dict objectForKey:@"curr_flow"]];
    }
    return ms;
}

-(void)saveRecordToContents:(NSDictionary *)dict
{
    NSInteger parentRecordId;
    NSInteger thisRecordId = [(NSNumber *)[dict valueForKey:@"RECORDID"] integerValue];
    NSString * levelName = [dict valueForKey:@"LEVELNAME"];
    NSUInteger level = NSNotFound;
    
    if (levelName == nil)
        return;
    
    if ([dict valueForKey:@"ctuse"] != nil)
    {
        NSString * contentText = [self makeContentTextFromRecord:[dict valueForKey:@"plain"]];
        NSString * simpleContentText = [self makeSimpleContentText:contentText];
        NSMutableDictionary * contentItem =  [[NSMutableDictionary alloc] initWithCapacity:6];
        
        [contentItem setValue:[dict valueForKey:@"id"] forKey:@"record"];
        [contentItem setValue:contentText forKey:@"text"];
        [contentItem setValue:simpleContentText forKey:@"simpletitle"];
        
        NSMutableArray * children = [self.contentTaggedItems valueForKey:[dict valueForKey:@"ctuse"]];
        [children addObject:contentItem];
    }
    
    level = [self getLevelIndex:levelName];
    
    
    if ([self checkSupressedContentLevel:level])
        return;
    
    // if no level defined, then this record is not part of contents
    if (level == NSNotFound || level >= kContStripMax)
        return;
    
    // writes last record to level
    for(NSUInteger i = level; i < kContStripMax; i++) {
        lastLevelRecord[i] = -1;
        contStrips[i] = nil;
    }
    
    lastLevelRecord[level] = thisRecordId;
    contStrips[level] = [[NSMutableDictionary alloc] init];
    [contStrips[level] setValue:[dict valueForKey:@"STPLAST"] forKey:@"STPLAST"];
    
    // gets parent for current record level
    parentRecordId = 0;
    for(NSInteger i = (NSInteger)level - 1; i >= 0; i--) {
        if (lastLevelRecord[i] >= 0) {
            parentRecordId = lastLevelRecord[i];
            break;
        }
    }
    
    
    NSString * contentText = [self makeContentTextFromRecord:[dict valueForKey:@"plain"]];
    NSString * simpleContentText = [self makeSimpleContentText:contentText];
    NSMutableDictionary * contentItem =  [[NSMutableDictionary alloc] initWithCapacity:7];
    
    [contentItem setValue:[NSNumber numberWithInteger:thisRecordId] forKey:@"record"];
    [contentItem setValue:[NSNumber numberWithInteger:parentRecordId] forKey:@"parent"];
    [contentItem setValue:contentText forKey:@"text"];
    [contentItem setValue:[NSNumber numberWithInteger:level] forKey:@"level"];
    [contentItem setValue:simpleContentText forKey:@"simpletitle"];
    NSString * strSubtext = [dict valueForKey:@"subtext"];
    if (strSubtext == nil)
        strSubtext = @"";
    NSMutableString * ms = [[NSMutableString alloc] initWithString:strSubtext];
    [contentItem setValue:ms forKey:@"subtext"];
    if ([dict valueForKey:@"ctdef"] != nil)
    {
        NSMutableArray * children = [[NSMutableArray alloc] init];
        [contentItem setValue:children forKey:@"children"];
        [self.contentTaggedItems setValue:children forKey:[dict valueForKey:@"ctdef"]];
    }
    
    [self.contentArray addObject:contentItem];
    
    self.lastContentItemInserted = contentItem;
    [contentItem setValue:[self.stpdefs objectForKey:levelName] forKey:@"STPDEF"];
    

    
    return;
}


-(NSString *)makeSimpleContentText:(NSString *)orig
{
    NSData * data = [[[orig lowercaseString] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] dataUsingEncoding:NSASCIIStringEncoding allowLossyConversion:YES];
    NSMutableString * str = [[NSMutableString alloc] init];
    const char * bytes = [data bytes];
    NSUInteger len = [data length];
    int lastType = 0;
    for (NSUInteger i = 0; i < len; i++)
    {
        unichar uc = bytes[i];
        if (isalpha(uc)) {
            if (lastType == 3) {
                [str replaceCharactersInRange:NSMakeRange([str length] - 1, 1) withString:@" "];
            }
            [str appendFormat:@"%C", uc];
            lastType = 1;
        } else if (isdigit(uc)) {
            [str appendFormat:@"%C", uc];
            lastType = 2;
        } else if (uc == '.') {
            if (lastType == 2) {
                [str appendString:@"."];
                lastType = 3;
            }
        } else if (uc == '-' || uc == 150) {
            if (lastType == 3) {
                [str replaceCharactersInRange:NSMakeRange([str length] - 1, 1) withString:@" "];
            } else if (lastType == 2) {
                [str appendString:@"-"];
                lastType = 4;
            } else if (lastType != 6) {
                [str appendString:@" "];
                lastType = 6;
            }
        } else if (uc == '\'') {
            [str appendString:@"'"];
            lastType = 5;
        } else if (uc == ' ') {
            if (lastType == 3) {
                [str replaceCharactersInRange:NSMakeRange([str length] - 1, 1) withString:@" "];
            } else if (![str hasSuffix:@" "] && [str length] > 0) {
                [str appendString:@" "];
            }
            lastType = 6;
        } else {
            if (lastType != 6) {
                [str appendString:@" "];
                lastType = 6;
            }
        }
    }
    
    while ([str hasSuffix:@" "]) {
        [str replaceCharactersInRange:NSMakeRange([str length] - 1, 1) withString:@""];
    }
    
    return str;
}

-(void)saveContents
{
    int ai = 0;
    int contentItemId = 0;
    for(NSMutableDictionary * contItem in self.contentArray)
    {
        contentItemId++;
        ai++;
        [contItem setValue:[NSNumber numberWithInt:contentItemId] forKey:@"itemid"];
        NSNumber * level = [contItem objectForKey:@"level"];
        NSNumber * parent = [contItem objectForKey:@"record"];
        if ([contItem valueForKey:@"children"] != nil)
        {
            NSSortDescriptor * sortDesc = [NSSortDescriptor sortDescriptorWithKey:@"simpletitle" ascending:YES selector:@selector(localizedCaseInsensitiveCompare:)];
            NSArray * children = [(NSArray *)[contItem valueForKey:@"children"] sortedArrayUsingDescriptors:[NSArray arrayWithObject:sortDesc]];
            
            for(NSMutableDictionary * contChild in children)
            {
                contentItemId++;
                [contChild setValue:[NSNumber numberWithInt:contentItemId] forKey:@"itemid"];
                [contChild setValue:parent forKey:@"parent"];
                [contChild setValue:[NSNumber numberWithInt:([level intValue] + 1)] forKey:@"level"];
            }
        }
        
        // writes to dump file
        fprintf(self->outputFile, "%d\t%d\t%d\t%s\t%s\t%s\n",
                [(NSNumber *)[contItem objectForKey:@"level"] intValue],
                [(NSNumber *)[contItem objectForKey:@"record"] intValue],
                [(NSNumber *)[contItem objectForKey:@"parent"] intValue],
                [[contItem objectForKey:@"text"] UTF8String],
                [[contItem objectForKey:@"simpletitle"] UTF8String],
                [[contItem objectForKey:@"subtext"] UTF8String]);
    }
    
    NSLog(@"Done - Saving Content");
}


@end

File: python/ContentsBuilder.py
import os
import os.path
from io import StringIO

import Flat
from FlatFileUtils import FlatFileTagString


kStackMax     = 64
kContStripMax = 240

class ContentLevelRecordMap:
    def __init__(self):
        self.reclev = [0] * kContStripMax
    def setEnd(self,endrec):
        self.reclev = [endrec] * kContStripMax
    def setRecordForLevel(self,rec,lev):
        for i in range(lev,len(self.reclev)):
            self.reclev[i] = rec
    def getRecordForLevel(self,lev):
        return self.reclev[lev]

class ContentsBuilder:
    def __init__(self):
        self.outputFile = None
        self.textFile = None
        self.contStrips = [None] * kContStripMax
        self.levelMapping = [0] * kContStripMax
        self.lastLevelRecord = [0] * kContStripMax
        self.outputDir = None
        self.inputFile = None
        self.levelFile = None
        self.levels = {}
        self.whiteSpaces = [' ', '\t', '\n', '\r']
        self.contentTaggedItems = {}
        self.stpdefs = {}
        self.contentArray = []
        self.lastContentItemInserted = {}
        self.maxRecordId = 0
        self.maxLevelAllowed = 7

    def validate(self):
        return len(self.inputFile) > 0 and len(self.outputDir) > 0 and len(self.levelFile) > 0

    def loadLevels(self):
        self.levels = {}
        with open(self.levelFile,'rt',encoding='utf-8') as rf:
            for line in rf:
                parts = line.strip().split('\t')
                if len(parts)==3:
                    self.levels[parts[2]] = int(parts[0])

    def process(self):
        self.loadLevels()

        count = 0

        self.outputFile = open(os.path.join(self.outputDir,"contents.txt"), "wt", encoding='utf-8')
        self.textFile = open(os.path.join(self.outputDir,"texts-b.txt"), "wt", encoding='utf-8')

        with open(self.inputFile, "rt",encoding='utf-8') as inputf:
            for line in inputf:
                part = line.strip().split('\t')
                while len(part) < 4:
                    part.append('')
                text = part[1].strip()
                levelName = part[2]
                styleName = part[3]
                recId = int(part[0])
                if self.maxRecordId<recId: self.maxRecordId = recId

                dict = {
                    'RECORDID': recId,
                    'LEVELNAME': levelName,
                    'STYLENAME': styleName,
                    'curr_flow': 'plain',
                    'plain': StringIO()
                }
                self.setDict(dict,text)

                print("{}\t{}\t{}\t{}".format(recId, dict['plain'].getvalue(), levelName, styleName), file=self.textFile)

                self.processText(dict)

                count +=1

                if count % 2000 == 0:
                    print(f'\rProcessed {count} lines', end='')
            print()

        self.setEndRecordIds()
        self.saveContents()
        self.outputFile.close()
        self.textFile.close()


    def processText(self,dict):
        text = dict['plain'].getvalue()
        recId = dict["RECORDID"]
        style = dict["STYLENAME"]
        level = dict["LEVELNAME"]

        if "STPDEF" in self.lastContentItemInserted:
            contentBuilding = self.lastContentItemInserted["STPDEF"]
            hook = contentBuilding[level]
            if not hook:
                hook = contentBuilding[style]
            if hook:
                plainText = Flat.makeContentTextFromRecord(text)
                if 'subtext' in self.lastContentItemInserted:
                    target = self.lastContentItemInserted["subtext"]
                    target.write("<STP:{}>{}".format(hook, plainText))
        # save record to contents
        return self.saveRecordToContents(dict)

    def getLevelIndex(self,levelName):
        if not levelName or len(levelName) == 0:
            return -1

        if levelName not in self.levels:
            return -1

        return self.levels[levelName]

    def checkSupressedContentLevel(self,nLevel):
        for i in range(kContStripMax - 1, -1, -1):
            if self.contStrips[i] and 'STPLAST' in self.contStrips[i] and self.contStrips[i]['STPLAST']:
                str = self.contStrips[i]["STPLAST"]
                level = self.getLevelIndex(str)
                # first STPLAST found is actually last STPLAST defined
                # so last STPLAST overrides all previously defined
                # therefore return after evaluating level index
                # :: current level (nLevel) must not be higher than level of found STPLAST
                # :: in order to write content item into contents
                return level < nLevel
        if nLevel>self.maxLevelAllowed: return True
        return False

    def setDict(self,dict,str):
        if not str: return

        start = 0
        end = 0
        status = 0
        foundTag = False
        foundChar = False
        foundCharSpec = False

        for i in range(len(str)):
            if status == 0:
                if str[i] == '<':
                    status = 1
                else:
                    foundChar = True
            elif status == 1:
                if str[i] == '<':
                    status = 0
                    foundCharSpec = True
                else:
                    start = i-1
                    status = 2
            elif status == 2:
                if str[i] == '>':
                    end = i
                    foundTag = True
                    status = 0
                elif str[i] == '"':
                    status = 3
            elif status == 3:
                if str[i] == '"':
                    status = 4
            elif status == 4:
                if str[i] == '"':
                    status = 3
                elif str[i] == '>':
                    end = i
                    foundTag = True
                    status = 0

            if foundTag:
                extractedTag = str[start:end+1]
                tag = FlatFileTagString()
                tag.clear()
                tag.appendString(extractedTag)
                foundTag = False
                arr = tag.createArray()

                if arr[0]=="FLOW":
                    dict['curr_flow'] = arr[2]
                elif arr[0]=="STPLAST":
                    if len(arr) == 1:
                        dict['STPLAST'] = ''
                    elif len(arr) == 3:
                        paraName = arr[2]
                        safeString = Flat.stringToSafe(paraName,"LE")
                        dict['STPLAST'] = safeString
                elif arr[0] == "CTDEF":
                    if len(arr) > 2:
                        dict['ctdef'] = arr[2]
                elif arr[0] == "CTUSE":
                    if len(arr) > 2:
                        dict['ctuse'] = arr[2]
                elif arr[0] == "STPDEF":
                    if len(arr) == 1:
                        self.stpdefs = {}
                    elif len(arr)==5:
                        if arr[4] in self.stpdefs:
                            del self.stpdefs[arr[4]]
                    elif len(arr) == 13:
                        # tagArr[2] must be equal to LV
                        # tagArr[10] must be equal to STP
                        targetString = Flat.stringToSafe(arr[4],'LE')
                        hookString = Flat.stringToSafe(arr[8],arr[6])
                        if targetString not in self.stpdefs:
                            self.stpdefs[targetString] = {hookString:arr[12]}
                        else:
                            self.stpdefs[targetString[hookString]] = arr[12]
                else:
                    ms = self.currentFlowText(dict)
                    ms.write(extractedTag)
            elif foundChar:
                ms = self.currentFlowText(dict)
                ms.write(str[i])
                foundChar = False
            elif foundCharSpec:
                ms = self.currentFlowText(dict)
                ms.write('<<')
                foundCharSpec = False

        return

    def currentFlowText(self,dict):
        cf = dict['curr_flow']
        if cf not in dict:
            dict[cf] = StringIO()
        return dict[cf]

    def saveRecordToContents(self,dict):
        thisRecordId = dict["RECORDID"]
        levelName = dict["LEVELNAME"]
        level = -1

        if not levelName: return

        if 'ctuse' in dict:
            contentText = Flat.makeContentTextFromRecord(dict['plain'],getvalue())
            simpleContentText = Flat.makeSimpleContentText(contentText)
            contentItem = {
                'record': dict['id'],
                'text': contentText,
                'simpletitle': simpleContentText
            }

            children = self.contentTaggedItems[dict['ctuse']]
            children.append(contentItem)
        level = self.getLevelIndex(levelName)

        if self.checkSupressedContentLevel(level): return

        # if no level defined, then this record is not part of contents
        if level < 0 or level >= kContStripMax: return

        # writes last record to level
        for i in range(level,kContStripMax):
            self.lastLevelRecord[i] = -1
            self.contStrips[i] = None

        self.lastLevelRecord[level] = thisRecordId
        stp = dict['STPLAST'] if 'STPLAST' in dict else ''
        self.contStrips[level] = {
            'STPLAST': stp
        }

        # gets parent for current record level
        parentRecordId = 0
        for i in range(level-1,-1,-1):
            if self.lastLevelRecord[i] >= 0:
                parentRecordId = self.lastLevelRecord[i]
                break

        contentText = Flat.makeContentTextFromRecord(dict['plain'].getvalue())
        simpleContentText = Flat.makeSimpleContentText(contentText)
        strSubtext = ''
        if 'subtext' in dict: strSubtext = dict["subtext"]
        subtext = StringIO(strSubtext)

        contentItem = {
            'record': thisRecordId,
            'parent': parentRecordId,
            'text': contentText,
            'level': level,
            'simpletitle': simpleContentText,
            'subtext': subtext
        }

        if 'ctdef' in dict:
            children = []
            contentItem['children'] = children
            self.contentTaggedItems[dict['ctdef']] = children

        self.contentArray.append(contentItem)
        self.lastContentItemInserted = contentItem
        if levelName in self.stpdefs:
            contentItem['STPDEF'] = self.stpdefs[levelName]

        return contentItem

    def setEndRecordIds(self):
        map = ContentLevelRecordMap()
        map.setEnd(self.maxRecordId)
        lastLevel = -1
        for cidx in range(len(self.contentArray)-1,-1,-1):
            ci = self.contentArray[cidx]
            ci['node_count'] = 0 if lastLevel <= ci['level'] else 1
            ci['nextSibling'] = map.getRecordForLevel(ci['level'])
            map.setRecordForLevel(ci['record'],ci['level'])
            lastLevel = ci['level']

    def saveContents(self):
        ai = 0
        contentItemId = 0
        for contItem in self.contentArray:
            contentItemId+=1
            ai+=1
            contItem['itemid'] = contentItemId
            level = contItem['level']
            parent = contItem['record']
            child_count = contItem['node_count']
            if 'children' in contItem:
                children = sorted(contItem['children'], key='simpletitle')
                for contChild in children:
                    contentItemId+=1
                    contChild['itemid'] = contentItemId
                    contChild['parent'] = parent
                    contChild['level'] = level + 1

            # writes to dump file
            self.outputFile.write("{}\t".format(contItem['level']))
            self.outputFile.write("{}\t".format(contItem['record']))
            self.outputFile.write("{}\t".format(contItem['parent']))
            self.outputFile.write("{}\t".format(contItem["text"]))
            self.outputFile.write("{}\t".format(contItem["simpletitle"]))
            self.outputFile.write("{}\t".format(contItem["subtext"].getvalue()))
            self.outputFile.write("{}\t".format(child_count))
            self.outputFile.write("{}\t".format('C'))
            node_type = 1 if child_count==0 else 2
            self.outputFile.write("{}\t".format(node_type))
            self.outputFile.write("{}\n".format(contItem['nextSibling']))

        print("Done - Saving Content")
